doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,481 @@
+
+* List of related documents gathered previously here https://hackmd.io/UVOtgj1wRZSsoNQCjkhq1g (TODO: Reduce to the most important ones):
+
+  * https://github.com/numpy/numpy/pull/12630
+
+    * Matti Picus NEP, discusses the technical side of subclassing  more from the side of ``ArrFunctions``
+
+  * https://hackmd.io/ok21UoAQQmOtSVk6keaJhw and https://hackmd.io/s/ryTFaOPHE
+
+    * (2019-04-30) Proposals for subclassing implementation approach.
+  
+  * Discussion about the calling convention of ufuncs and need for more powerful UFuncs: https://github.com/numpy/numpy/issues/12518
+
+  * 2018-11-30 developer meeting notes: https://github.com/BIDS-numpy/docs/blob/master/meetings/2018-11-30-dev-meeting.md and subsequent draft for an NEP: https://hackmd.io/6YmDt_PgSVORRNRxHyPaNQ
+
+    * BIDS Meeting on November 30, 2018 and document by Stephan Hoyer about what numpy should provide and thoughts of how to get there. Meeting with Eric Wieser, Matti Pincus, Charles Harris, Tyler Reddy, Stéfan van der Walt, and Travis Oliphant. (TODO: Full list of people?)
+    * Important summaries of use cases.
+
+  * SciPy 2018 brainstorming session: https://github.com/numpy/numpy/wiki/Dtype-Brainstorming
+
+    * Good list of user stories/use cases.
+    * Lists some requirements and some ideas on implementations
+
+
+
+References and Footnotes
+------------------------
+
+.. _pandas_extension_arrays: https://pandas.pydata.org/pandas-docs/stable/development/extending.html#extension-types
+
+.. _xarray_dtype_issue: https://github.com/pydata/xarray/issues/1262
+
+.. _pygeos: https://github.com/caspervdw/pygeos
+
+.. _new_sort: https://github.com/numpy/numpy/pull/12945
+
+.. _gh-12518: https://github.com/numpy/numpy/issues/12518
+
+.. _value_based: Value based promotion denotes the behaviour that NumPy will inspect the value of scalars (and 0 dimensional arrays) to decide what the output dtype should be. ``np.array(1)`` typically gives an "int64" array, but ``np.array([1], dtype="int8") + 1`` will retain the "int8" of the first array.
+
+.. _safe_casting: Safe casting denotes the concept that the value held by one dtype can be represented by another one without loss/change of information. Within current NumPy there are two slightly different usages. First, casting to string is considered safe, although it is not safe from a type perspective (it is safe in the sense that it cannot fail); this behaviour should be considered legacy. Second, int64 is considered to cast safely to float64 even though float64 cannot represent all int64 values correctly.
+
+.. _flexible_dtype: A flexible dtype is a dtype for which conversion is not always safely possible. This is for example the case for current string dtypes, which can have different lengths. It is also true for datetime64 due to its attached unit. A non-flexible dtype should typically have a canonical representation (i.e. a float64 may be in non-native byteorder, but the default is native byte order).
```

Comment:
This definition is not consistent with `PyArray_ISFLEXIBLE`, which includes only void, str, and bytes.